### PR TITLE
feat: implemented cow swapper for exchange client

### LINKED ISSUE #4838: Implement cowswapper for `Swapper2` interface
### Overview

Implement cowswapper for `Swapper2` interface

### References and additional details

See `src/lib/swapper/swappers/LifiSwapper/LifiSwapper2.ts` and `src/lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper2.ts`

### Acceptance Criteria

Can perform trades on this swapper exactly as in production.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

Half day

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Implements cow swapper for new exchange client. Includes several fixes, commented inline.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #4838

## Risk

Low risk of something borked with cow swapper.

## Testing

- Test user-facing cow swap is working
- Test user-facing cow quotes are still correct

### Engineering

Test the new cow swapper with multi hop flag on

### Operations

See testing notes above. Should be no change from prod.

## Screenshots (if applicable)

New cow swapper executing trades:
image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4882** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4882" target="_blank">Graphite</a> 
    * **PR #4884** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4884" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4884?utm_source=stack-comment).

File: src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts [hunk 1]
```diff
@@ -186,5 +186,5 @@ export const getInputOutputRatioFromQuote = ({
     .plus(sellSideNetworkFeeUsdPrecision)
     .plus(sellSideProtocolFeeUsdPrecision)
 
-  return netSendAmountUsdPrecision.div(netReceiveAmountUsdPrecision).toNumber()
```
Comment (woodenfurniture) [inline src/state/apis/swappers/helpers/getInputOutputRatioFromQuote.ts:189]: fixed incorrect quote ratio calculation such that bad quotes were ranked higher than good ones

File: src/lib/swapper/swappers/LifiSwapper/endpoints.ts [hunk 2]
```diff
@@ -78,11 +76,19 @@ export const lifiApi: Swapper2Api = {
...
+    const lifiRoute = tradeQuoteMetadata.get(quoteId)
+    if (!lifiRoute) throw Error(`missing trade quote metadata for quoteId ${quoteId}`)
+
+    const getStatusRequest: GetStatusRequest = {
+      txHash,
+      bridge: lifiRoute.steps[0].tool,
+      fromChain: lifiRoute.fromChainId,
+      toChain: lifiRoute.toChainId,
+    }
+
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/LifiSwapper/endpoints.ts:91]: fixes error during status polling for lifi trades

File: src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts [hunk 3]
```diff
@@ -103,9 +103,7 @@ export async function getCowSwapTradeQuote(
   })
 
   // don't show buy amount if less than min sell amount
-  const isSellAmountBelowMinimum = bnOrZero(normalizedSellAmountCryptoBaseUnit).lt(
-    minimumCryptoBaseUnit,
-  )
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/CowSwapper/getCowSwapTradeQuote/getCowSwapTradeQuote.ts:108]: fixes cow swapper returning an amount when the input is 0. The logic here was including the hardcoded minimum when it should be using the inputted amount

File: src/state/slices/tradeQuoteSlice/selectors.ts [hunk 4]
```diff
@@ -45,10 +45,11 @@ export const selectActiveSwapperApiResponse: Selector<ReduxState, ApiQuote | und
     selectActiveSwapperName,
     (quotes, activeSwapperName) => {
       const selectedQuote = quotes.find(quote => quote.swapperName === activeSwapperName)
-      if (selectedQuote) {
+      if (selectedQuote?.quote !== undefined) {
```
Comment (woodenfurniture) [inline src/state/slices/tradeQuoteSlice/selectors.ts:48]: dont return the selected quote if it's in error, this fix prevents the UI showing error status when valid quotes are available

File: src/state/slices/tradeQuoteSlice/selectors.ts [hunk 5]
```diff
@@ -45,10 +45,11 @@ export const selectActiveSwapperApiResponse: Selector<ReduxState, ApiQuote | und
...
         return selectedQuote
       } else {
-        return quotes.length > 0 ? quotes[0] : undefined
+        const successfulQuotes = quotes.filter(({ quote }) => quote !== undefined)
+        return successfulQuotes.length > 0 ? successfulQuotes[0] : undefined
```
Comment (woodenfurniture) [inline src/state/slices/tradeQuoteSlice/selectors.ts:52]: default to the first SUCCESSFUL quote, this fix prevents defaulting to a quote that forces the UI to be stuck in error state

File: src/lib/swapper/api.ts [hunk 6]
```diff
@@ -288,15 +288,21 @@ export type GetUnsignedTxArgs = {
   feeAssetUsdRate: string
 } & FromOrXpub
 
+// the client should never need to know anything about this payload, and since it varies from
+// swapper to swapper, the type is declared this way to prevent generics hell while ensuring the
+// data originates from the correct place (assuming no casting).
+export type UnsignedTx2 = Nominal<object, 'UnsignedTx2'>
```
Comment (woodenfurniture) [inline src/lib/swapper/api.ts:294]: this change is key to allow all swappers to conform to the same interface with entering generics hell. Feel pretty strongly about keeping this as an opaque type instead of a generic interface

File: src/components/Trade/TradeConfirm/AssetToAsset.tsx [hunk 7]
```diff
@@ -21,42 +22,42 @@ export const AssetToAsset: FC<AssetToAssetProps> = ({
...
-      case TxStatus.Failed:
+      case status === TxStatus.Failed:
         return (
           <Circle bg={red} size='100%'>
             <CloseIcon p={1} />
           </Circle>
         )
-      case TxStatus.Pending:
+      // when the trade is submitting, treat unknown status as pending so the spinner spins
+      case status === TxStatus.Pending || (isSubmitting && status === TxStatus.Unknown):
```
Comment (woodenfurniture) [inline src/components/Trade/TradeConfirm/AssetToAsset.tsx:47]: not a user facing change, but allows the new UI to show the spinner while a trade is executing for unknown status. This prevents cowswap from sitting there doing nothing without a spinner and suddenly jumping to green tick without a spinner - in other words prevents user thinking app is broken when they sign the tx.

File: src/components/MultiHopTrade/hooks/useTradeExecution/useTradeExecution.ts [hunk 8]
```diff
@@ -118,19 +121,20 @@ export const useTradeExecution = ({
       },
     )
 
-    const sellTxId = await swapper.executeTrade({
+    const sellTxHash = await swapper.executeTrade({
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useTradeExecution/useTradeExecution.ts:124]: Disregard my comment on the previous PR since this is now tackled here

File: src/lib/swapper/api.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/lib/swapper/api.ts:294]: Can we make this a `Record<string, any>` instead of an object?
The `object` type in TS is an old construct from [TS 2.2](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#object-type) which represents anything that's not a primitive (i.e an "object" at the very low "everything is an object" level, e.g a function, an array...)

File: src/lib/swapper/swappers/CowSwapper/endpoints.ts [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+      return { id, receiveAddress, affiliateBps, ...tradeQuote }
+    })
+  },
+
+  getUnsignedTx: async ({ from, tradeQuote, stepIndex }: GetUnsignedTxArgs): Promise<CowSignTx> => {
+    const { accountNumber, buyAsset, sellAsset, sellAmountBeforeFeesCryptoBaseUnit } =
+      tradeQuote.steps[stepIndex]
+    const { receiveAddress } = tradeQuote
+
+    const buyToken = !isNativeEvmAsset(buyAsset.assetId)
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/CowSwapper/endpoints.ts:69]: This is copied and pasted from the previous implementation but can we use the `address` vernacular here? 🙏🏽

File: src/lib/swapper/swappers/LifiSwapper/endpoints.ts [hunk 10]
```diff
@@ -78,11 +76,19 @@ export const lifiApi: Swapper2Api = {
   },
 
   checkTradeStatus: async ({
-    tradeId,
+    quoteId,
+    txHash,
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/LifiSwapper/endpoints.ts:80]: Feels like this should belong in a separate PR since this is not CoW-related, probably in #4882, though since this is related to the new swapper under flag, happy to keep it here

**Review by gomesalexandre — APPROVED** [review]
LGTM - happy to get this one in as a stack

Regression tested CoW with the multi-hop flag off:

### Input

image

### Preview

image

### Pending trade

image

### Confirmed state

image


Tested CoW with the flag on:

### Input

image


### Preview

image

### Pending trade

image


### Confirmed state

image

File: src/lib/swapper/swappers/LifiSwapper/endpoints.ts (hunk 10 above)
Comment (woodenfurniture) [reply src/lib/swapper/swappers/LifiSwapper/endpoints.ts:80]: thank you ser

File: src/lib/swapper/api.ts (hunk 6 above)
Comment (woodenfurniture) [reply src/lib/swapper/api.ts]: 4e8a1196a9684a0524b90a32228ce31e9cd66962
